gh-61460: Stronger HMAC in multiprocessing

tiran opened the pull request:
See also https://github.com/python/cpython/issues/61460

Signed-off-by: Christian Heimes <christian@python.org>

https://github.com/python/cpython/issues/61460 formerly known as <!-- issue-number: [bpo-17258](https://bugs.python.org/issue17258) --> https://bugs.python.org/issue17258 <!-- /issue-number -->


<!-- gh-issue-number: gh-61460 -->
* Issue: gh-61460
<!-- /gh-issue-number -->

tiran committed:
bpo-17258: Stronger HMAC in multiprocessing

Signed-off-by: Christian Heimes <christian@python.org>

gpshead commented on Misc/NEWS.d/next/Library/2020-05-25-12-42-36.bpo-17258.lf2554.rst:
```diff
@@ -0,0 +1 @@
+:mod:`multiprocessing` supports stronger HMAC algorithms
```
Explain what they're used for as it is an internal implementation detail that they get used.

gpshead commented on Lib/multiprocessing/connection.py:
```diff
@@ -734,30 +735,93 @@ def PipeClient(address):
 
-def deliver_challenge(connection, authkey):
+_mac_algo_re = re.compile(
+    rb'^{(?P<digestmod>(md5|sha256|sha384|sha3_256|sha3_384))}'
```
I wouldn't bother including the 384 variants, nothing will use them.

gpshead commented on Lib/multiprocessing/connection.py:
```diff
@@ -734,30 +735,93 @@ def PipeClient(address):
+    return True
+
+
+def deliver_challenge(connection, authkey, digestmod=None):
```
I suggest defaulting to "sha256".  There's no point to this code if we don't use something more modern.

gpshead reviewed (commented):
Why jump through all the hoops to specify the digest in the protocol? Don't we always control both ends of the connection so there should never be a situation where negotiation and understanding of what was used is needed?

That'd be a lot less complicated.

And not prone to the potential problem this has of always stooping to the lowest level decided upon out by the challenge initiator rather than _requiring_ a specific hash to be used on the channel.

pitrou commented on Lib/multiprocessing/connection.py:
```diff
@@ -734,30 +735,93 @@ def PipeClient(address):
-def deliver_challenge(connection, authkey):
+_mac_algo_re = re.compile(
+    rb'^{(?P<digestmod>(md5|sha256|sha384|sha3_256|sha3_384))}'
+    rb'(?P<payload>.*)$'
```
How long can the payload be? Perhaps it's overkill (and potentially costly) to make this a specific capture group instead of simply taking the rest of the string after `digestmod`.

gpshead commented on Lib/multiprocessing/connection.py:
```diff
@@ -734,30 +735,93 @@ def PipeClient(address):
+        # old-style MD5 with fallback
+        digestmod = None
+
+    if digestmod is None:
```
DONE: rather than keying solely off of the prefix match, we can check the length of message is > MESSAGE_LENGTH only then check for the {algorithm} prefix.  prefix bytes will occur naturally within some deliver_challenge based md5 messages given how short our prefixes are.  we don't want to misinterpret those.

gpshead commented on Lib/multiprocessing/connection.py:
```diff
@@ -734,30 +735,93 @@ def PipeClient(address):
+    # message: {digest}payload, the MAC protects header and payload
+    mo = _mac_algo_re.match(message)
+    if mo is not None:
+        digestmod = mo.group('digestmod').decode('ascii')
```
DONE: digestmod is an old term, they haven't been modules in ages.  digest_name makes more sense.

gpshead committed:
Merge branch 'main' into bpo17258-multiproc-md5

gpshead committed:
Reworked to be more robust while keeping the idea.

The protocol modification idea remains, but we now take advantage of the
message length as an indicator of legacy vs modern protocol version.  No
more regular expression usage.  We now default to HMAC-SHA256, but do so
in a way that will be compatible when communicating with older clients
or older servers. No protocol transition period is needed.

More unittests to verify these claims remain true are required.

gpshead committed:
further update to the comment to match impl.

gpshead committed:
Merge branch 'main' into bpo17258-multiproc-md5
